### PR TITLE
[docs] Update docs on code-review process

### PR DESCRIPTION
**Pull request by banach-space**
Clarify what to do in cases where multiple reviewers commented on a
patch, but only one reviewer explicitly accepted.

This is mostly to standardize what the expectations are as folks tend to
approach this differently.

### CODE REVIEW

**Comment by banach-space** [comment]
Follow-on from a discussion here: https://github.com/llvm/llvm-project/pull/111533

My assumption is that this is documenting what many of us already practice, so hopefully this is nothing particularly new. Please let me know if you feel otherwise.

Any suggestion who else to include in the review?

File: llvm/docs/Contributing.rst
Comment (kuhar) [inline llvm/docs/Contributing.rst:110]: Should we s/LGTM/approval/ (as in the button in the github ui)? I think 'LGTM' might be the language coupled to some older iteration of our code review process that feels a bit out of place now

**Review by nikic — CHANGES_REQUESTED** [review]
This document is just an overview, we have detailed documentation of the code review policy in https://llvm.org/docs/CodeReview.html, which is also already linked. 

In particular, https://llvm.org/docs/CodeReview.html#lgtm-how-a-patch-is-accepted explains that it is the responsibility of the approving reviewer to consider other feedback on the patch and indicate if their approval is conditional.

File: llvm/docs/Contributing.rst
Comment (joker-eph) [inline llvm/docs/Contributing.rst:119]: Instead of adding more text here, I think we should trim it down and link to the dedicated page: https://llvm.org/docs/CodeReview.html

Right now you're evolving the text in a direction that contradicts what this other page is saying I believe.

Comment (banach-space) [reply llvm/docs/Contributing.rst]: > I think we should trim it down and link to the dedicated page

Agreed, I think that repeating similar information in multiple places might be confusing. I went ahead and deleted a paragraph from Contributing.rst to avoid that.

File: llvm/docs/CodeReview.rst
Comment (joker-eph) [inline llvm/docs/CodeReview.rst:163]: The second part of the sentence is a bit strongly worded, what do you mean to cover here?

**Review by joker-eph — APPROVED** [review]
LG overall, I have one comment though, thanks!

File: llvm/docs/CodeReview.rst
Comment (kuhar) [reply llvm/docs/CodeReview.rst:163]: +1, I don't think we want to effectively require a re-approval after every push.
For example, when a new reviewer comes and add a typo/nit comment, I'd expect that the author can fix that and merge based on the previous approval given how insignificant the comment (and subsequent fix) is. I think this is a standard practice today.

Comment (banach-space) [reply llvm/docs/CodeReview.rst]: I've expanded this paragraph. @Jakub, hopefully that meets your criteria (which I mostly agree with).

> what do you mean to cover here?

What should happen when there's a "LGTM", but then another reviewer leaves new comments afterwards? I wanted to clarify that the original "LGTM" wouldn't apply to the new comments (unless, like Jakub points out, those would be  basic nits etc).

> For example, when a new reviewer comes and add a typo/nit comment, I'd expect that the author can fix that and merge based on the previous approval given how insignificant the comment (and subsequent fix) is. 

This is assuming that there won't be more substantial comments following the nits - is it a valid assumption? I guess the reviewer should make that clear, e.g. "Just some nits, no need to wait for my approval".

File: llvm/docs/CodeReview.rst
Comment (kuhar) [inline llvm/docs/CodeReview.rst:165]: To follow the style above

```suggestion
Exceptions apply, e.g., there's no need to confirm that a comment requesting a
```

File: llvm/docs/CodeReview.rst
Comment (kuhar) [inline llvm/docs/CodeReview.rst:166]: I'd make it less specific to allow for some wiggle room we've traditionally had

```suggestion
minor issue (like a typo or other 'nit') to be fixed has been addressed (this should be evident from the code).
```

File: llvm/docs/CodeReview.rst
Comment (banach-space) [reply llvm/docs/CodeReview.rst]: >  but I don't see how "all new changes have to be reviewed " is addressing what you describe here ("another reviewer leaves new comments afterwards").

That's been updated since :)

File: llvm/docs/CodeReview.rst
Comment (preames) [inline llvm/docs/CodeReview.rst:161]: This advise talks about action the reviewer should take, that seems backwards here.  Unless I'm misreading, the problematic situation here is a prior LGTM followed by changes by a *different* reviewer and the author landing the patch without addressing those changes.

I'd suggest something along the lines of the following:
Only approval from a single reviewer is required; however the patch author is always responsible for making sure that all feedback on a review has been properly addressed or responded to.  If additional feedback is provided after acceptance (by the same reviewer or another), the author should use their best judgement in deciding whether that feedback can be incorporated into the change without comment (say a typo) or requires further review discussion.  Similarly, the patch author is responsible for incorporating post commit feedback on the change.  

I'd put this paragraph just above the one talking about unqualified LGTM

Comment (banach-space) [reply llvm/docs/CodeReview.rst:161]: >  Unless I'm misreading, the problematic situation here is a prior LGTM followed by comments by a different reviewer and the author landing the patch without addressing those comments.

Yes

>  (edit: for typo fix)

No, for more substantial changes. For "typos", I added an exception.

>  If additional feedback is provided after acceptance (by the same reviewer or another), the author should use their best judgement in deciding whether that feedback can be incorporated into the change without comment (say a typo) or requires further review discussion.

To me this sounds that any comments "post LGTM" are "second category" (i.e. "for the author to decide whether to address them or not" vs "need to be addressed"). I'm suggesting that we should treat such comments as any other comments prior to LGTM (i.e., "need to be addressed"). I feel that that otherwise we are discouraging people from reading/commenting on patches post LGTM.

Comment (preames) [reply llvm/docs/CodeReview.rst:161]: > > Unless I'm misreading, the problematic situation here is a prior LGTM followed by comments by a different reviewer and the author landing the patch without addressing those comments.
> 
> Yes
> 
> > (edit: for typo fix)
> 
> No, for more substantial changes. For "typos", I added an exception.

(You're responding to a note I made about editting my text to fix a typo...)
 
> > If additional feedback is provided after acceptance (by the same reviewer or another), the author should use their best judgement in deciding whether that feedback can be incorporated into the change without comment (say a typo) or requires further review discussion.
> 
> To me this sounds that any comments "post LGTM" are "second category" (i.e. "for the author to decide whether to address them or not" vs "need to be addressed"). I'm suggesting that we should treat such comments as any other comments prior to LGTM (i.e., "need to be addressed"). I feel that that otherwise we are discouraging people from reading/commenting on patches post LGTM.

If you have suggested wording, feel free to suggest.  I don't read this the way you do, particular following the prior sentence that you trimmed.

Comment (banach-space) [reply llvm/docs/CodeReview.rst:161]: Hi @preames , thanks for that suggestion and sorry for the delay - I had to step away and re-think this a bit. I do want to incorporate your feedback, but also want to make sure we avoid duplication and keep this document succinct. So I trimmed your suggestion a bit as per the comments below (hope this won't come across as nit-picking).

**Point 1:**
> [From Phillip's suggestion]
> Only approval from a single reviewer is required; 

This point is already mentioned earlier in the document: https://github.com/llvm/llvm-project/blob/c2063de1593610eda0f4de33c3b89324642ed54c/llvm/docs/CodeReview.rst?plain=1#L148-L150
 
**Point 2:**
> [From Phillip's suggestion]
> however the patch author is always responsible for making sure that all feedback on a review has been properly addressed or responded to. 

Isn't this a contradicting one existing policy?
https://github.com/llvm/llvm-project/blob/c2063de1593610eda0f4de33c3b89324642ed54c/llvm/docs/CodeReview.rst?plain=1#L152-L159

Specifically:

> it is the responsibility of the reviewer to have reviewed all of the discussion and feedback from all reviewers ensuring that all feedback has been addressed and that all other reviewers will almost surely be satisfied with the patch being approved.

I'd just leave this out.

**Point 3:**

> [From Phillip's suggestion]
> Similarly, the patch author is responsible for incorporating post commit feedback on the change.

This is already covered here: https://github.com/llvm/llvm-project/blob/c2063de1593610eda0f4de33c3b89324642ed54c/llvm/docs/CodeReview.rst?plain=1#L46-L49

**Point 4:**

Which leaves this bit from Phillip's original suggestion (emphasis mine):
> [From Phillip's suggestion]
If additional feedback is provided after acceptance (by the same reviewer or another), **the author should use their best judgement** in deciding whether that feedback can be incorporated into the change without comment (say a typo) or requires further review discussion. 

This a bit softer requirement compared to what I added "these need to be addressed" and sounds more along the lines of what @kuhar would prefer? Let me use that, but let me add small clarification at the end. Let me know what you think.

File: llvm/docs/Contributing.rst
Comment (nikic) [inline llvm/docs/Contributing.rst:107]: I feel like this has cut out too much. E.g. I think the "ping" convention should still be mentioned here. This is quite relevant to new contributors, who otherwise don't need to read the full scope of the code review documentation.

Comment (banach-space) [reply llvm/docs/Contributing.rst:107]: I can leave it if you feel strongly about it, but IMO this is duplicating information and that can be confusing. Sometimes less is more and also - why not _encourage_ people to read our policy instead? it's an important document and it's there for a reason.

File: llvm/docs/CodeReview.rst
Comment (preames) [reply llvm/docs/CodeReview.rst]: You have seemingly managed to complete miss the spirit of my original suggestion, and are misreading the current policy in my view.  If the document implies that the author is not responsible for incorporating feedback (because that's only the reviewers responsibility), the document is simply wrong.  I don't read it that way, but if you do, stop and fix that wording before working on this smaller item.

File: llvm/docs/Contributing.rst
Comment (banach-space) [reply llvm/docs/Contributing.rst]: I'm keen to progress this, so I've just restored that note. Thanks!

File: llvm/docs/CodeReview.rst
Comment (banach-space) [reply llvm/docs/CodeReview.rst]: > You have seemingly managed to complete miss the spirit of my original suggestion

That's not intentional. Would you be able to share a diff so that I better understand what you had in mind? 

> If the document implies that the author is not responsible for incorporating feedback

I don’t believe the document implies that. I think it’s fairly clear that:

1. The author is responsible for addressing/incorporating the feedback, which is already covered under "Acknowledge All Reviewer Feedback" (so perhaps we can leave that as is).
2. The reviewer, when posting an unqualified LGTM, is responsible for ensuring that all feedback has been addressed (i.e., “I've reviewed all comments and confirm they’ve been addressed").

The existing wording covers those points well. The part that is unclear IMO is what happens **after LGTM** and **before commit** when new comments are posted. Hopefully, the additional paragraph I proposed clarifies that.

**Comment by banach-space** [comment]
Gentle ping :)

**Comment by joker-eph** [comment]
Still LGTM, but we should hear back from @nikic / @preames.

File: llvm/docs/CodeReview.rst
Comment (kuhar) [inline llvm/docs/CodeReview.rst:164]: Let's keep this consistent with the style used above
```suggestion
requires further review discussion. More substantial comments (e.g., about the
```

**Review by kuhar — APPROVED** [review]
LGTM % nit

**Comment by banach-space** [comment]
Ping @nikic and @preames - are you OK with me landing this in its current form? Thanks!

**Review by nikic — APPROVED** [review]
LGTM

**Review by dcaballe — APPROVED** [review]

**Comment by banach-space** [comment]
Hey folks, thank you for all the reviews! If there are no new comments, I will land this tomorrow.

**Pull request merged**
